engine/crates/operation-checks/src/aggregate_field_usage.rs

```diff
@@ -0,0 +1,421 @@
+mod async_graphql;
+
+use std::collections::HashMap;
+
+#[derive(Debug)]
+pub struct FieldUsage<'a> {
+    pub increment: u64,
+    /// field id -> usage count
+    pub counts: &'a mut HashMap<usize, u64>,
+}
+
+impl FieldUsage<'_> {
+    pub fn with_increment(self, new_increment: u64) -> Self {
+        Self {
+            increment: new_increment,
+            counts: self.counts,
+        }
+    }
+
+    fn register(&mut self, field_id: usize) {
+        if let Some(count) = self.counts.get_mut(&field_id) {
+            *count += self.increment;
+        } else {
+            self.counts.insert(field_id, self.increment);
+        }
+    }
+}
+
+/// (type name, field name) -> field type
+#[derive(Debug)]
+pub struct Schema {
+    pub fields: Vec<(String, String, String)>,
+    pub query_type_name: String,
+    pub mutation_type_name: String,
+    pub subscription_type_name: String,
+}
+
+impl Schema {
+    fn find_field(&self, type_name: &str, field_name: &str) -> Option<usize> {
+        self.fields
+            .binary_search_by_key(&(type_name, field_name), |(type_name, field_name, _)| {
+                (type_name, field_name)
+            })
+            .ok()
+    }
+}
+
+#[derive(Debug)]
+pub struct Query {
+    pub fragments: HashMap<String, (String, usize)>,
+    pub operation_type: OperationType,
+    pub root_selection: usize,
+    pub selections: Vec<(usize, Selection)>,
+}
+
+#[derive(Debug)]
+pub enum OperationType {
+    Query,
+    Mutation,
+    Subscription,
+}
+
+#[derive(Debug)]
+pub enum Selection {
+    Field(String, Option<usize>),
+    FragmentSpread(String),
+    InlineFragment(String, usize),
+}
+
+pub fn aggregate_field_usage(query: &Query, schema: &Schema, usage: &mut FieldUsage<'_>) {
+    let ty = match query.operation_type {
+        OperationType::Query => &schema.query_type_name,
+        OperationType::Mutation => &schema.mutation_type_name,
+        OperationType::Subscription => &schema.subscription_type_name,
+    };
+    aggregate_field_usage_inner(query.root_selection, ty, query, schema, usage);
+}
+
+fn aggregate_field_usage_inner(
+    selection_id: usize,
+    ty: &str,
```

Comment:
Might be clearer as `r#type`

Reply:
renamed to parent_type in upcoming commit